Migrate WCS Stripe Oauth code to Stripe

chickenn00dle opened the pull request:
This is a work in progress for moving WCS Stripe Connect Oauth code into Stripe.

### Changes proposed in this Pull Request:

Add the following from WCS:

- Connect api
- Base REST controller
- REST deauthorize controller
- REST oauth-init controller
- REST oauth-connect controller

I've also removed the Jetpack dependency from Oauth requests, so both Jetpack and WCS are no longer necessary for the OAuth flow to work.

### Testing

The changes in this PR are dependent upon the following two PRs in [woocommerce-admin](https://github.com/woocommerce/woocommerce-admin/pull/4933) and [woocommerce-connect-server](https://github.com/Automattic/woocommerce-connect-server/pull/1443)

To test these changes now, you will need to have both of the above installed/running locally. Otherwise, you can wait for the Connect Server PR to be merged as the changes there won't impact the current OAuth flow.

To test:

- Get a fresh install of WP set up on a test site and install WooCommerce
- With the WC-Admin changes above checked out, kick off the setup wizard
- Skip the WCS/Jetpack install step
- Complete the Stripe Payment task. With these changes, the oauth flow should work here even though Jetpack and WCS are not installed.
- Go to the Stripe settings page
- Look for the **Reset Stripe account keys** button and click it: https://d.pr/i/njErdY
- The page should reload and your keys should no longer be saved.
- Look for the **Setup or link an existing Stripe Account** button and click it: https://d.pr/i/G2fAkt
- This should kick off the OAuth flow again and re-save your Stripe keys

Note: When testing with a sandbox Stripe account, you will need to manually change instances of publishable_key and secret_key to test_publishable_key and test_secret_key in the following WC-Admin files:

    client/task-list/tasks/payments/stripe.js
    client/task-list/tasks/payments/methods.js


### Questions

I wasn't sure about a few pieces of the OAuth code so just copied these over anyway to be safe.

- Is adding "useful data" to the OAuth request necessary in `WC_Stripe_Connect_API`?
- Are the deauth controller and related methods necessary? How are these used and are they still relevant now that Jetpack is no longer a requirement?

-------------------
- [x] Make sure your changes respect [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).

Closes #1253

chickenn00dle commented:
All of the OAuth functionality from WCS has been moved over to Stripe now. All that is left is to remove the Jetpack dependency.

Before doing that, I'll first need to submit PRs to the following repos:

- [ ] Connect Server: Remove Jetpack dependency
- [ ] WC-Admin: Remove check for WCS and Jetpack during Stripe onboarding step

chickenn00dle commented:
An important note, this should not be shipped until after Stripe Connect changes. See pb0GrA-Pd-p2 for more details.

dechov commented:
As a potentially simpler alternative to the AJAX requests, the URL could be generated dynamically using ordinary links (such as [these](https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/blob/c4a6be036346646d1d7ea2c6d0bd3f1f249f9121/includes/settings/settings-ppec.php#L17-L38)) that are redirected (in [this way](https://github.com/Automattic/woocommerce-payments/blob/44b79cf5eba9a3c00ae22a64e59f5104cb9f3547/includes/class-wc-payments-account.php#L393-L394)).

> Are the deauth controller and related methods necessary? How are these used and are they still relevant now that Jetpack is no longer a requirement?

The `/deauthorize` endpoint won't work without authentication, since the site's ID is (was) used as a lookup for the connected Stripe account. Not sure if that's an issue, except that it might break [this UI](https://github.com/Automattic/wp-calypso/blob/e72794c5ec674c25992cd95afb2c555314bf5bb1/client/extensions/woocommerce/app/settings/payments/stripe/payment-method-stripe-connect-account.js#L95-L101) (from Store on WP.com and also loaded into the Stripe settings screen via WooCommerce Services).

chickenn00dle commented:
Thanks for the review @dechov 

> As a potentially simpler alternative to the AJAX requests, the URL could be generated dynamically using ordinary links (such as these) that are redirected (in this way).

That is a much simpler approach! I've made the change in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1267/commits/4102d0bb9cea060d738693e2cb549ffe0a1eb3ea

> The /deauthorize endpoint won't work without authentication, since the site's ID is (was) used as a lookup for the connected Stripe account. Not sure if that's an issue, except that it might break this UI 

Thanks for clarifying. Since this endpoint will no longer be functional given we've removed authentication, I've removed this controller in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1267/commits/65b195fd2212f0788fb2fd3fa6c592c13e57dd23

As for the UI you pointed out in calypso, I'm not familiar with how this interacts and so I'm not certain what further changes are necessary. Will we need to remove that bit in calypso, or accommodate the functionality in Stripe? I'll spend some more time digging into this next week if you're not sure.

chickenn00dle commented:
I pushed https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1267/commits/ea1ba8c5c2bc017aa28162666ae923bacecc2e92 to address some errors that were introduced by the previous commit. This is ready for review again!

dechov commented:
> As for the UI you pointed out in calypso, I'm not familiar with how this interacts and so I'm not certain what further changes are necessary. Will we need to remove that bit in calypso, or accommodate the functionality in Stripe?

Good question, which _may_ warrant some further attention on the server. _Getting account info_ and _de-authorizing account_ are features of the existing integration that are broken or unavailable with latest server changes, and I realize now we could (I think) have just encoded the state in a proper way to securely identify the correct account (just as the OAuth connection flow does) and persisted it in the plugin. Do you see a way to do this without much extra complexity? If not, it _might_ not be worth prioritizing – but we should keep in mind that for accounts connecting without that, we won't have a way of identifying on our side again, for either showing info or deauthorization (or any hypothetical other purpose).

I'll try and get to some testing and further review on this before the week is out.

dechov commented:
Tested from just this plugin, and the connection worked well:

Before OAuth | After OAuth
-- | --
<img width="736" alt="before-oauth" src="https://user-images.githubusercontent.com/1867547/95907812-2433c880-0d6a-11eb-988f-15849bfd96ec.png"> | <img width="731" alt="after-oauth" src="https://user-images.githubusercontent.com/1867547/95907837-2bf36d00-0d6a-11eb-9a0c-59daa807da3c.png">

I was seeing some odd behavior in a case where the settings had never been properly saved (with the "Save changes" button at bottom of the settings screen):

When I went to a different page (e.g. "Plugins"), the admin notice "Stripe is almost ready. To get started, […]" was showing up. Didn't actually try the front-end to see if the gateway showed / worked, but the notice behaved normally after clicking "Save changes" – even after resetting the keys, connecting again, etc.

I think this _might_ have to do with the [default settings code](https://github.com/Automattic/woocommerce-services/blob/788057f99853c41e1fa46402be35dccaca555384/classes/class-wc-connect-stripe.php#L229-L243), which wasn't copied over?

Also, should we have an inline ToS acceptance (such as in https://github.com/woocommerce/woocommerce-admin/commit/95265111eb1143fffbd276f2603ede897d093362) here as well?

chickenn00dle commented:
Thanks for testing @dechov!

> When I went to a different page (e.g. "Plugins"), the admin notice "Stripe is almost ready. To get started, […]" was showing up. Didn't actually try the front-end to see if the gateway showed / worked, but the notice behaved normally after clicking "Save changes" – even after resetting the keys, connecting again, etc.

I am not able to reproduce this behavior. The notice appears only when Stripe is not connected for me. Here is a screencast of what I am seeing: https://d.pr/i/2oA8CU

Could you share the steps you are taking for this to happen, or a screencast if possible?

> I think this might have to do with the default settings code, which wasn't copied over?

Hmm. The notice appears to be displayed if `secret_key` is empty, and we are not on the Stripe settings page:
https://github.com/woocommerce/woocommerce-gateway-stripe/blob/34ee0dbc8b9804e0220e744a06abb8d133121912/includes/admin/class-wc-stripe-admin-notices.php#L167

Other Stripe settings shouldn't impact this behavior.

As for not copying over this method, I actually meant to remove the `$default_options` line altogether as I didn't find it necessary to fetch default values now that we are handling the connection flow from within the Stripe plugin. Are there unintended consequences of going this route that I've missed?

> Also, should we have an inline ToS acceptance (such as in woocommerce/woocommerce-admin@9526511) here as well?

Good call. I've added this in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1267/commits/34ee0dbc8b9804e0220e744a06abb8d133121912

dechov commented:
I can reproduce testing on a fresh install by going through the connection flow following the directions above (or via Stripe settings) and then navigating to a different admin screen. The issue appears to be traceable to [this check](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/52761eaedf058a26a5e6e317b6cfa1de74f71211/includes/class-wc-stripe-api.php#L41), which assumes that test and live key settings would both be _set_ even if empty and/or unused. (The Settings form submits default values for all blank fields – not totally sure if there could be any other unintended consequences of not having all of the fields set, or if we can just fix this one check.)

chickenn00dle commented:
Thanks for digging into this @dechov!

I've copied over the default settings code in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1267/commits/b0ca244b03d510eebeff05ccf06ca90e71203747

allendav commented:
@chickenn00dle - looks like https://github.com/Automattic/woocommerce-connect-server/pull/1443 was merged a few weeks ago. Which PR needs to land next? This one or https://github.com/woocommerce/woocommerce-admin/pull/4933 ?

If the latter, will we have to wait for that to be merged to WooCommerce core too?

We will have a mix of Stripe (and WooCommerce) plugin versions in the wild. Will that be a problem? What if users have the wc-admin plugin installed?

cc @bmccotter @c-shultz

chickenn00dle commented:
Hey @allendav 👋

> Which PR needs to land next? This one or woocommerce/woocommerce-admin#4933 ?

This one would be next. You can find more details about this here: pb0GrA-Pd-p2 

> We will have a mix of Stripe (and WooCommerce) plugin versions in the wild. Will that be a problem? What if users have the wc-admin plugin installed?

Based on my testing, this shouldn't be a problem. At least, assuming we ship in the order mentioned in the P2 post.

WC Admins' role in stripe onboarding is to require Jetpack and WCS to be installed and active before attempting to render the OAuth components during the onboarding payment task. So the only real issue I foresee are cases in which a merchant has WC/WC Admin updated to include the changes from woocommerce/woocommerce-admin#4933 but an outdated version of Stripe for some reason. I imagine this is unlikely given part of WC onboarding is to install Stripe, but assuming there are instances like these, onboarding will fall back to manual config.

Here is a breakdown of the possible mixtures of Stripe/WC versions:

old Stripe and old WC/admin - JP and WCS required
old Stripe and new WC/admin - covered above
new Stripe and old WC/admin - JP and WCS required
new Stripe and new WC/admin - this is what we want

As I mentioned, I did some testing to see what order we need to ship in, but as far as I can tell, I've been the only one to do this. So it would be great to have another person to confirm my findings.

I hope this answers your questions @allendav 

Let me know if anything else needs to be clarified.

allendav commented on includes/connect/class-wc-stripe-connect.php:
```diff
@@ -0,0 +1,185 @@
+				wp_safe_redirect( remove_query_arg( array( 'wcs_stripe_state', 'wcs_stripe_code' ) ) );
+				exit;
+			// redirect from credentials reset
+			} elseif ( isset( $_GET['reset_stripe_api_credentials'] ) ) {
```
It isn't readily apparent to me this is protected by a nonce. Is it?

chickenn00dle replied:
Good catch! I missed this. I've added one here: https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1267/commits/3e8b0e59183b9f70122a1c27c47d23f10849c1c0

allendav commented on includes/admin/stripe-settings.php:
```diff
@@ -3,6 +3,33 @@
+	);
+
+	$api_credentials_text = sprintf(
+		__( '%1$sReset Stripe account keys%2$s', 'woocommerce-gateway-stripe' ),
```
Let's make it clearer what this will cause to happen. Could we change the linked text to

Clear all Stripe account keys. This will disable any connection to Stripe.

chickenn00dle replied:
Sure thing! Changed in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1267/commits/f9ad9f38bba43311f5d8998705e2b71325202034

allendav commented on includes/admin/stripe-settings.php:
```diff
@@ -3,6 +3,33 @@
+
+	if ( ! is_wp_error( $oauth_url ) ) {
+		$api_credentials_text = sprintf(
+			__( '%1$sSetup or link an existing Stripe Account%2$s By clicking this button you agree to the %3$sTerms of Service%2$s. Or manually enter Stripe keys below.', 'woocommerce-gateway-stripe' ),
```
Please make account lowercase.
Please add a period after it or the closing tag.
Please add a comma after Or
Please replace Stripe keys with Stripe account keys for clarity.

chickenn00dle replied:
Also addressed in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1267/commits/f9ad9f38bba43311f5d8998705e2b71325202034

allendav commented on includes/admin/stripe-settings.php:
```diff
@@ -3,6 +3,33 @@
+
+		);
+	} else {
+		$api_credentials_text = __( 'Manually enter Stripe keys below.', 'woocommerce-gateway-stripe' );
```
Please replace Stripe keys with Stripe account keys for clarity.

chickenn00dle replied:
https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1267/commits/f9ad9f38bba43311f5d8998705e2b71325202034

allendav commented on includes/admin/stripe-settings.php:
```diff
@@ -33,6 +60,11 @@
 			'description' => $this->display_admin_settings_webhook_description(),
 		),
+		'api_credentials'               => array(
+			'title'       => __( 'API Credentials', 'woocommerce-gateway-stripe' ),
```
Could we change API Credentials to "Stripe account keys" ?

chickenn00dle replied:
https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1267/commits/f9ad9f38bba43311f5d8998705e2b71325202034

allendav commented on includes/connect/class-wc-stripe-connect-api.php:
```diff
@@ -0,0 +1,244 @@
+			$business_data['business_name'] = html_entity_decode( get_bloginfo( 'name' ), ENT_QUOTES );
+			$business_data['first_name']    = $current_user->user_firstname;
+			$business_data['last_name']     = $current_user->user_lastname;
+			$business_data['phone']         = '';
```
WooCommerce core, which is likely active at this point, may have a phone number for the current user in user meta.

allendav commented on includes/connect/class-wc-stripe-connect-api.php:
```diff
@@ -0,0 +1,244 @@
+
+			$wc_countries = WC()->countries;
+
+			if ( method_exists( $wc_countries, 'get_base_address' ) ) {
```
Related, no action here: https://github.com/woocommerce/woocommerce-gateway-stripe/issues/1341

allendav commented on includes/connect/class-wc-stripe-connect-api.php:
```diff
@@ -0,0 +1,244 @@
+			}
+
+			$http_timeout = 60; // 1 minute
+			if ( function_exists( 'wc_set_time_limit' ) ) {
```
Probably don't need to check for this? wc_set_time_limit has been around since 2.6 and this plugin requires 3.0

chickenn00dle replied:
Good call. Removed this check in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1267/commits/c286591a251eda4439044105aaff161011b1fba0

allendav commented on includes/connect/class-wc-stripe-connect-api.php:
```diff
@@ -0,0 +1,244 @@
+						)
+					);
+				}
+				return $response;
```
What happens if we exit this way? Seems bad.

chickenn00dle replied:
Good catch! I'm not sure why this was returned originally, but this should be an error. I've addressed this in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1267/commits/c18c5527253345618ac522a6a3a34ddb6d2ceb3d

allendav commented on includes/connect/class-wc-stripe-connect-rest-controller.php:
```diff
@@ -0,0 +1,172 @@
+	/**
+	 * Stripe Connect base REST controller class.
+	 */
+	class WC_Stripe_Connect_REST_Controller extends WP_REST_Controller {
```
Should this class be abstract? e.g. rest_base only appears to be defined in the descendants below

chickenn00dle replied:
Made this abstract in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1267/commits/de0d448f1d4901e53adc21927403ef9ec0ab0f1b

allendav commented on includes/connect/class-wc-stripe-connect.php:
```diff
@@ -0,0 +1,185 @@
+
+			// While we are at it, let's also clear the account_id and
+			// test_account_id if present.
+			unset( $options['account_id'] );
```
Should these unsets be within the conditionals? To avoid blowing away both account IDs? i.e. unset test_account_id if in testmode, unset the live account_id if not.

chickenn00dle replied:
Truthfully, I'm not certain. It seems to me that we are wanting both of these values to be removed if they are present, but to be safe, I've moved the unsets to within the above conditional logic in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1267/commits/4c66a0801c9d8bbcb645bb3aec3646eecfd92e25

allendav commented on includes/connect/class-wc-stripe-connect.php:
```diff
@@ -0,0 +1,185 @@
+			$options = get_option( self::SETTINGS_OPTION, array() );
+
+			if ( isset( $options['testmode'] ) && 'yes' === $options['testmode'] ) {
+				return isset( $options['test_publishable_key'], $options['test_secret_key'] ) && $options['test_publishable_key'] && $options['test_secret_key'];
```
This will probably work, but I'd feel more confident if we trimmed before testing for falsiness.

chickenn00dle replied:
I had assumed these values would have any whitespace removed by the time they are set. But further trimming can't hurt! Added in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1267/commits/4c66a0801c9d8bbcb645bb3aec3646eecfd92e25

allendav commented on includes/connect/class-wc-stripe-connect.php:
```diff
@@ -0,0 +1,185 @@
+			if ( isset( $options['testmode'] ) && 'yes' === $options['testmode'] ) {
+				return isset( $options['test_publishable_key'], $options['test_secret_key'] ) && $options['test_publishable_key'] && $options['test_secret_key'];
+			} else {
+				return isset( $options['publishable_key'], $options['secret_key'] ) && $options['publishable_key'] && $options['secret_key'];
```
Same comment as above

chickenn00dle replied:
https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1267/commits/4c66a0801c9d8bbcb645bb3aec3646eecfd92e25

allendav commented on woocommerce-gateway-stripe.php:
```diff
@@ -315,8 +326,45 @@ public function add_emails( $email_classes ) {
+		return;
+	}
+
+	wc_stripe();
```
Neat pattern.

allendav commented:
If we don't allow test keys to be obtained from the connect server (is that correct?) then

- we shouldn't show the `Setup or link an existing Stripe account` button when in test mode
- we should hide the `Setup or link an existing Stripe account` button if the user checks Enable Test Mode (and show it if they uncheck after checking it)

I say all this because during testing it seemed I was always getting live keys back

allendav commented:
If the user starts editing things and THEN clicks the "Setup or link" button, they'll lose their edits if they proceed. Please consider disabling the button if they start making changes. This seems more likely than not since the "Setup or link" button is not the first thing on the page.

![leave](https://user-images.githubusercontent.com/1595739/96787525-bf4e2300-13a6-11eb-8a61-331e41ea643c.gif)

chickenn00dle commented:
Thanks for the review @allendav!

I've addressed the inline comments above. 

> If we don't allow test keys to be obtained from the connect server (is that correct?) 

We do allow test keys to be obtained from connect server: https://d.pr/i/4cVSPO

> Please consider disabling the button if they start making changes. 

I've done this in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1267/commits/3965195f36b17685903d8ec10d93f7992f8d468c

chickenn00dle commented:
Resolved conflict. This is ready for review again!

allendav reviewed (approved):
Tested successfully with staging and production - got keys for both. I do think we could continue to refine the user experience, and will open an issue for that, but this can go in now.

You can merge when you're ready and we'll take care of the rest.
